// firebase-admin/lib/utils/validator.js

"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.isBuffer = isBuffer;
exports.isArray = isArray;
exports.isNonEmptyArray = isNonEmptyArray;
exports.isBoolean = isBoolean;
exports.isNumber = isNumber;
exports.isString = isString;
exports.isBase64String = isBase64String;
exports.isNonEmptyString = isNonEmptyString;
exports.isObject = isObject;
exports.isNonNullObject = isNonNullObject;
exports.isUid = isUid;
exports.isPassword = isPassword;
exports.isEmail = isEmail;
exports.isPhoneNumber = isPhoneNumber;
exports.isISODateString = isISODateString;
exports.isUTCDateString = isUTCDateString;
exports.isURL = isURL;
exports.isTopic = isTopic;
exports.isTaskId = isTaskId;

function isBuffer(value) {
    return value instanceof Buffer;
}

function isArray(value) {
    return Array.isArray(value);
}

function isNonEmptyArray(value) {
    return isArray(value) && value.length !== 0;
}

function isBoolean(value) {
    return typeof value === 'boolean';
}

function isNumber(value) {
    return typeof value === 'number' && !isNaN(value);
}

function isString(value) {
    return typeof value === 'string';
}

function isBase64String(value) {
    if (!isString(value)) {
        return false;
    }
    return /^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/.test(value);
}

function isNonEmptyString(value) {
    return isString(value) && value !== '';
}

function isObject(value) {
    return typeof value === 'object' && !isArray(value);
}

function isNonNullObject(value) {
    return isObject(value) && value !== null;
}

function isUid(uid) {
    return typeof uid === 'string' && uid.length > 0 && uid.length <= 128;
}

function isPassword(password) {
    
    return typeof password === 'string' && password.length >= 6;
}

function isEmail(email) {
    if (typeof email !== 'string') {
        return false;
    }
    
    const re = /^[^@]+@[^@]+$/;
    return re.test(email);
}

function isPhoneNumber(phoneNumber) {
    if (typeof phoneNumber !== 'string') {
        return false;
    }

    const re1 = /^\+/;
    
    const re2 = /[\da-zA-Z]+/;
    return re1.test(phoneNumber) && re2.test(phoneNumber);
}

function isISODateString(dateString) {
    try {
        return isNonEmptyString(dateString) &&
            (new Date(dateString).toISOString() === dateString);
    }
    catch (e) {
        return false;
    }
}

function isUTCDateString(dateString) {
    try {
        return isNonEmptyString(dateString) &&
            (new Date(dateString).toUTCString() === dateString);
    }
    catch (e) {
        return false;
    }
}

function isURL(urlStr) {
    if (typeof urlStr !== 'string') {
        return false;
    }
    
    const re = /[^a-z0-9:/?#[\]@!$&'()*+,;=.\-_~%]/i;
    if (re.test(urlStr)) {
        return false;
    }
    try {
        const uri = new URL(urlStr);
        const scheme = uri.protocol;
        if (scheme !== 'http:' && scheme !== 'https:') {
            return false;
        }
        const hostname = uri.hostname;
        // Validate hostname strictly to match previous behavior and prevent weak/invalid domains.
        // Must be alphanumeric with optional dashes/underscores, separated by dots.
        // Cannot start/end with dot or dash (mostly).
        // This regex is safe (no nested quantifiers with overlap).
        if (!/^[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?(\.[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?)*$/.test(hostname)) {
            // Check for IPv6 literals which are valid but behave differently.
            // Node 'new URL' keeps brackets for IPv6: [::1] -> [::1]
            // Check for IPv6 address (simple check for brackets)
            if (!/^\[[a-fA-F0-9:.]+\]$/.test(hostname)) {
                return false;
            }
        }
        // Restore strict pathname validation: (/chars+)*/?
        // Where chars can be a combination of: a-z A-Z 0-9 - _ . ~ ! $ & ' ( ) * + , ; = : @ %
        const pathnameRe = /^(\/[\w\-.~!$'()*+,;=:@%]+)*\/?$/;
        // Validate pathname.
        const pathname = uri.pathname;
        if (pathname &&
            pathname !== '/' &&
            !pathnameRe.test(pathname)) {
            return false;
        }
        return true;
    }
    catch (e) {
        return false;
    }
}
/**
 * Validates that the provided topic is a valid FCM topic name.
 *
 * @param topic - The topic to validate.
 * @returns Whether the provided topic is a valid FCM topic name.
 */
function isTopic(topic) {
    if (typeof topic !== 'string') {
        return false;
    }
    const VALID_TOPIC_REGEX = /^(\/topics\/)?(private\/)?[a-zA-Z0-9-_.~%]+$/;
    return VALID_TOPIC_REGEX.test(topic);
}
/**
 * Validates that the provided string can be used as a task ID
 * for Cloud Tasks.
 *
 * @param taskId - the task ID to validate.
 * @returns Whether the provided task ID is valid.
 */
function isTaskId(taskId) {
    if (typeof taskId !== 'string') {
        return false;
    }
    const VALID_TASK_ID_REGEX = /^[A-Za-z0-9_-]+$/;
    return VALID_TASK_ID_REGEX.test(taskId);
}
