// google-auth-library/build/src/gtoken/getCredentials.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.getCredentials = getCredentials;
const path = require("path");
const fs = require("fs");
const util_1 = require("util");
const errorWithCode_1 = require("./errorWithCode");
const readFile = fs.readFile
    ? (0, util_1.promisify)(fs.readFile)
    : async () => {
        
        throw new errorWithCode_1.ErrorWithCode('use key rather than keyFile.', 'MISSING_CREDENTIALS');
    };
var ExtensionFiles;
(function (ExtensionFiles) {
    ExtensionFiles["JSON"] = ".json";
    ExtensionFiles["DER"] = ".der";
    ExtensionFiles["CRT"] = ".crt";
    ExtensionFiles["PEM"] = ".pem";
    ExtensionFiles["P12"] = ".p12";
    ExtensionFiles["PFX"] = ".pfx";
})(ExtensionFiles || (ExtensionFiles = {}));

class JsonCredentialsProvider {
    keyFilePath;
    constructor(keyFilePath) {
        this.keyFilePath = keyFilePath;
    }
    
    async getCredentials() {
        const key = await readFile(this.keyFilePath, 'utf8');
        let body;
        try {
            body = JSON.parse(key);
        }
        catch (error) {
            const err = error;
            throw new Error(`Invalid JSON key file: ${err.message}`);
        }
        const privateKey = body.private_key;
        const clientEmail = body.client_email;
        if (!privateKey || !clientEmail) {
            throw new errorWithCode_1.ErrorWithCode('private_key and client_email are required.', 'MISSING_CREDENTIALS');
        }
        return { privateKey, clientEmail };
    }
}

class PemCredentialsProvider {
    keyFilePath;
    constructor(keyFilePath) {
        this.keyFilePath = keyFilePath;
    }
    
    async getCredentials() {
        const privateKey = await readFile(this.keyFilePath, 'utf8');
        return { privateKey };
    }
}

class P12CredentialsProvider {
    
    async getCredentials() {
        throw new errorWithCode_1.ErrorWithCode('*.p12 certificates are not supported after v6.1.2. ' +
            'Consider utilizing *.json format or converting *.p12 to *.pem using the OpenSSL CLI.', 'UNKNOWN_CERTIFICATE_TYPE');
    }
}

class CredentialsProviderFactory {
    
    static create(keyFilePath) {
        const keyFileExtension = path.extname(keyFilePath);
        switch (keyFileExtension) {
            case ExtensionFiles.JSON:
                return new JsonCredentialsProvider(keyFilePath);
            case ExtensionFiles.DER:
            case ExtensionFiles.CRT:
            case ExtensionFiles.PEM:
                return new PemCredentialsProvider(keyFilePath);
            case ExtensionFiles.P12:
            case ExtensionFiles.PFX:
                return new P12CredentialsProvider();
            default:
                throw new errorWithCode_1.ErrorWithCode('Unknown certificate type. Type is determined based on file extension. ' +
                    'Current supported extensions are *.json, and *.pem.', 'UNKNOWN_CERTIFICATE_TYPE');
        }
    }
}

async function getCredentials(keyFilePath) {
    const provider = CredentialsProviderFactory.create(keyFilePath);
    return provider.getCredentials();
}
