// google-auth-library/build/src/gtoken/getToken.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.getToken = getToken;
const jwsSign_1 = require("./jwsSign");

const GOOGLE_TOKEN_URL = 'https://oauth2.googleapis.com/token';

const GOOGLE_GRANT_TYPE = 'urn:ietf:params:oauth:grant-type:jwt-bearer';

const generateRequestOptions = (tokenOptions) => {
    return {
        method: 'POST',
        url: GOOGLE_TOKEN_URL,
        data: new URLSearchParams({
            grant_type: GOOGLE_GRANT_TYPE, 
            assertion: (0, jwsSign_1.getJwsSign)(tokenOptions),
        }),
        responseType: 'json',
        retryConfig: {
            httpMethodsToRetry: ['POST'],
        },
    };
};

async function getToken(tokenOptions) {
    if (!tokenOptions.transporter) {
        throw new Error('No transporter set.');
    }
    try {
        const gaxiosOptions = generateRequestOptions(tokenOptions);
        const response = await tokenOptions.transporter.request(gaxiosOptions);
        return response.data;
    }
    catch (e) {

        const err = e;
        const errorData = err.response?.data;
        if (errorData?.error) {
            err.message = `${errorData.error}: ${errorData.error_description}`;
        }
        throw err;
    }
}
