// google-auth-library/build/src/gtoken/googleToken.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.GoogleToken = void 0;
const gaxios_1 = require("gaxios");
const tokenHandler_1 = require("./tokenHandler");
const revokeToken_1 = require("./revokeToken");

class GoogleToken {
    
    tokenOptions;
    
    tokenHandler;
    
    constructor(options) {
        this.tokenOptions = options || {};
        
        this.tokenOptions.transporter = this.tokenOptions.transporter || {
            request: opts => (0, gaxios_1.request)(opts),
        };
        if (!this.tokenOptions.iss) {
            this.tokenOptions.iss = this.tokenOptions.email;
        }
        if (typeof this.tokenOptions.scope === 'object') {
            this.tokenOptions.scope = this.tokenOptions.scope.join(' ');
        }
        this.tokenHandler = new tokenHandler_1.TokenHandler(this.tokenOptions);
    }
    get expiresAt() {
        return this.tokenHandler.tokenExpiresAt;
    }
    
    get accessToken() {
        return this.tokenHandler.token?.access_token;
    }
    
    get idToken() {
        return this.tokenHandler.token?.id_token;
    }
    
    get tokenType() {
        return this.tokenHandler.token?.token_type;
    }
    
    get refreshToken() {
        return this.tokenHandler.token?.refresh_token;
    }
    
    hasExpired() {
        return this.tokenHandler.hasExpired();
    }
    
    isTokenExpiring() {
        return this.tokenHandler.isTokenExpiring();
    }
    getToken(callbackOrOptions, opts = { forceRefresh: false }) {
        
        let callback;
        if (typeof callbackOrOptions === 'function') {
            callback = callbackOrOptions;
        }
        else if (typeof callbackOrOptions === 'object') {
            opts = callbackOrOptions;
        }
        
        const promise = this.tokenHandler.getToken(opts.forceRefresh ?? false);
        
        if (callback) {
            promise.then(token => callback(null, token), callback);
        }
        return promise;
    }
    revokeToken(callback) {
        if (!this.accessToken) {
            return Promise.reject(new Error('No token to revoke.'));
        }
        const promise = (0, revokeToken_1.revokeToken)(this.accessToken, this.tokenOptions.transporter);
        
        if (callback) {
            promise.then(() => callback(), callback);
        }
        
        this.tokenHandler = new tokenHandler_1.TokenHandler(this.tokenOptions);
    }
    
    get googleTokenOptions() {
        return this.tokenOptions;
    }
}
exports.GoogleToken = GoogleToken;
