// google-auth-library/build/src/gtoken/tokenHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenHandler = void 0;
const getToken_1 = require("./getToken");
const getCredentials_1 = require("./getCredentials");

class TokenHandler {
    
    token;
    
    tokenExpiresAt;
    
    inFlightRequest;
    tokenOptions;
    
    constructor(tokenOptions) {
        this.tokenOptions = tokenOptions;
    }
    
    async processCredentials() {
        if (!this.tokenOptions.key && !this.tokenOptions.keyFile) {
            throw new Error('No key or keyFile set.');
        }
        if (!this.tokenOptions.key && this.tokenOptions.keyFile) {
            const credentials = await (0, getCredentials_1.getCredentials)(this.tokenOptions.keyFile);
            this.tokenOptions.key = credentials.privateKey;
            this.tokenOptions.email = credentials.clientEmail;
        }
    }
    
    isTokenExpiring() {
        if (!this.token || !this.tokenExpiresAt) {
            return true;
        }
        const now = new Date().getTime();
        const eagerRefreshThresholdMillis = this.tokenOptions.eagerRefreshThresholdMillis ?? 0;
        return this.tokenExpiresAt <= now + eagerRefreshThresholdMillis;
    }
    
    hasExpired() {
        const now = new Date().getTime();
        if (this.token && this.tokenExpiresAt) {
            const now = new Date().getTime();
            return now >= this.tokenExpiresAt;
        }
        return true;
    }
    
    async getToken(forceRefresh) {
        
        await this.processCredentials();
        
        if (this.inFlightRequest && !forceRefresh) {
            return this.inFlightRequest;
        }
        
        if (this.token && !this.isTokenExpiring() && !forceRefresh) {
            return this.token;
        }
        
        try {
            this.inFlightRequest = (0, getToken_1.getToken)(this.tokenOptions);
            const token = await this.inFlightRequest;
            
            this.token = token;
            this.tokenExpiresAt =
                new Date().getTime() + (token.expires_in ?? 0) * 1000;
            return token;
        }
        finally {
            
            this.inFlightRequest = undefined;
        }
    }
}
exports.TokenHandler = TokenHandler;
