// firebase-admin/lib/phone-number-verification/phone-number-verification-api-client-internal.js

"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.FirebasePhoneNumberVerificationError = exports.FPNV_ERROR_CODE_MAPPING = exports.FPNV_TOKEN_INFO = exports.JWKS_URL = void 0;
const error_1 = require("../utils/error");
exports.JWKS_URL = 'https://fpnv.googleapis.com/v1beta/jwks';
exports.FPNV_TOKEN_INFO = {
    url: 'https://firebase.google.com/docs/phone-number-verification',
    verifyApiName: 'verifyToken()',
    jwtName: 'Firebase Phone Verification token',
    shortName: 'FPNV token',
    typ: 'JWT',
};
exports.FPNV_ERROR_CODE_MAPPING = {
    INVALID_ARGUMENT: 'invalid-argument',
    INVALID_TOKEN: 'invalid-token',
    EXPIRED_TOKEN: 'expired-token',
};

class FirebasePhoneNumberVerificationError extends error_1.PrefixedFirebaseError {
    constructor(code, message) {
        super('phone-number-verification', code, message);

        this.__proto__ = FirebasePhoneNumberVerificationError.prototype;
    }
}
exports.FirebasePhoneNumberVerificationError = FirebasePhoneNumberVerificationError;
