// google-auth-library/build/src/util.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.LRUCache = void 0;
exports.snakeToCamel = snakeToCamel;
exports.originalOrCamelOptions = originalOrCamelOptions;
exports.removeUndefinedValuesInObject = removeUndefinedValuesInObject;
exports.isValidFile = isValidFile;
exports.getWellKnownCertificateConfigFileLocation = getWellKnownCertificateConfigFileLocation;
const fs = require("fs");
const os = require("os");
const path = require("path");
const WELL_KNOWN_CERTIFICATE_CONFIG_FILE = 'certificate_config.json';
const CLOUDSDK_CONFIG_DIRECTORY = 'gcloud';

function snakeToCamel(str) {
    return str.replace(/([_][^_])/g, match => match.slice(1).toUpperCase());
}

function originalOrCamelOptions(obj) {
    
    function get(key) {
        const o = (obj || {});
        return o[key] ?? o[snakeToCamel(key)];
    }
    return { get };
}

class LRUCache {
    capacity;
    
    #cache = new Map();
    maxAge;
    constructor(options) {
        this.capacity = options.capacity;
        this.maxAge = options.maxAge;
    }
    
    #moveToEnd(key, value) {
        this.#cache.delete(key);
        this.#cache.set(key, {
            value,
            lastAccessed: Date.now(),
        });
    }
    
    set(key, value) {
        this.#moveToEnd(key, value);
        this.#evict();
    }
    
    get(key) {
        const item = this.#cache.get(key);
        if (!item)
            return;
        this.#moveToEnd(key, item.value);
        this.#evict();
        return item.value;
    }
    
    #evict() {
        const cutoffDate = this.maxAge ? Date.now() - this.maxAge : 0;
        
        let oldestItem = this.#cache.entries().next();
        while (!oldestItem.done &&
            (this.#cache.size > this.capacity || 
                oldestItem.value[1].lastAccessed < cutoffDate) 
        ) {
            this.#cache.delete(oldestItem.value[0]);
            oldestItem = this.#cache.entries().next();
        }
    }
}
exports.LRUCache = LRUCache;

function removeUndefinedValuesInObject(object) {
    Object.entries(object).forEach(([key, value]) => {
        if (value === undefined || value === 'undefined') {
            delete object[key];
        }
    });
    return object;
}

async function isValidFile(filePath) {
    try {
        const stats = await fs.promises.lstat(filePath);
        return stats.isFile();
    }
    catch (e) {
        return false;
    }
}

function getWellKnownCertificateConfigFileLocation() {
    const configDir = process.env.CLOUDSDK_CONFIG ||
        (_isWindows()
            ? path.join(process.env.APPDATA || '', CLOUDSDK_CONFIG_DIRECTORY)
            : path.join(process.env.HOME || '', '.config', CLOUDSDK_CONFIG_DIRECTORY));
    return path.join(configDir, WELL_KNOWN_CERTIFICATE_CONFIG_FILE);
}

function _isWindows() {
    return os.platform().startsWith('win');
}
