// fast-xml-parser/src/v6/OutputBuilders/ParserOptionsBuilder.js
import trimParser from "../valueParsers/trim.js";
import booleanParser from "../valueParsers/booleanParser.js";
import currencyParser from "../valueParsers/currency.js";
import numberParser from "../valueParsers/number.js";

const defaultOptions = {
  nameFor: {
    text: "#text",
    comment: "",
    cdata: "",
  },

  piTag: false,
  declaration: false, 
  tags: {
    valueParsers: [

    ]
  },
  attributes: {
    prefix: "@_",
    suffix: "",
    groupBy: "",

    valueParsers: [

    ]
  },
  dataType: {

  }
}

const withJoin = ["trim", "join", "number", "boolean", "currency"]
const withoutJoin = ["trim", "number", "boolean", "currency"]

export function buildOptions(options) {
  
  const finalOptions = { ...defaultOptions };

  finalOptions.tags.valueParsers.push(...withJoin)
  if (!this.preserveOrder)
    finalOptions.tags.valueParsers.push(...withoutJoin);

  finalOptions.attributes.valueParsers.push(...withJoin)

  copyProperties(finalOptions, options);
  return finalOptions;
}

function copyProperties(target, source) {
  for (let key in source) {
    if (Object.prototype.hasOwnProperty.call(source, key)) {
      if (typeof source[key] === 'object' && !Array.isArray(source[key])) {
        
        if (typeof target[key] === 'undefined') {
          target[key] = {};
        }
        copyProperties(target[key], source[key]);
      } else {
        
        target[key] = source[key];
      }
    }
  }
}

export function registerCommonValueParsers(options) {
  return {
    "trim": new trimParser(),
    
    "boolean": new booleanParser(),
    "number": new numberParser({
      hex: true,
      leadingZeros: true,
      eNotation: true
    }),
    "currency": new currencyParser(),
    
  }
}
