// fast-xml-parser/src/v6/OptionsBuilder.js

import { JsObjOutputBuilder } from './OutputBuilders/JsObjBuilder.js';

export const defaultOptions = {
  preserveOrder: false,
  removeNSPrefix: false, 
  
  stopNodes: [], 
  
  htmlEntities: false,
  
  tags: {
    unpaired: [],
    nameFor: {
      cdata: false,
      comment: false,
      text: '#text'
    },
    separateTextProperty: false,
  },
  attributes: {
    ignore: false,
    booleanType: true,
    entities: true,
  },

  only: [], 
  hierarchy: false, 
  skip: [], 

  select: [], 
  stop: [], 
  OutputBuilder: new JsObjOutputBuilder(),
};

export const buildOptions = function (options) {
  const finalOptions = { ...defaultOptions };
  copyProperties(finalOptions, options)
  return finalOptions;
};

function copyProperties(target, source) {
  for (let key in source) {
    if (Object.prototype.hasOwnProperty.call(source, key)) {
      if (key === 'OutputBuilder') {
        target[key] = source[key];
      } else if (typeof source[key] === 'object' && !Array.isArray(source[key])) {
        
        if (typeof target[key] === 'undefined') {
          target[key] = {};
        }
        copyProperties(target[key], source[key]);
      } else {
        
        target[key] = source[key];
      }
    }
  }
}
