// firebase-functions/lib/params/index.js
const require_rolldown_runtime = require('../_virtual/rolldown_runtime.js');
const require_params_types = require('./types.js');

var params_exports =  require_rolldown_runtime.__export({
	BUCKET_PICKER: () => require_params_types.BUCKET_PICKER,
	Expression: () => require_params_types.Expression,
	clearParams: () => clearParams,
	databaseURL: () => databaseURL,
	declaredParams: () => declaredParams,
	defineBoolean: () => defineBoolean,
	defineFloat: () => defineFloat,
	defineInt: () => defineInt,
	defineJsonSecret: () => defineJsonSecret,
	defineList: () => defineList,
	defineSecret: () => defineSecret,
	defineString: () => defineString,
	expr: () => expr,
	gcloudProject: () => gcloudProject,
	multiSelect: () => require_params_types.multiSelect,
	projectID: () => projectID,
	select: () => require_params_types.select,
	storageBucket: () => storageBucket
});

const majorVersion = typeof "7" !== "undefined" ? "7" : "0";
const GLOBAL_SYMBOL = Symbol.for(`firebase-functions:params:declaredParams:v${majorVersion}`);
const globalSymbols = globalThis;
if (!globalSymbols[GLOBAL_SYMBOL]) {
	globalSymbols[GLOBAL_SYMBOL] = [];
}
const declaredParams = globalSymbols[GLOBAL_SYMBOL];

function registerParam(param) {
	for (let i = 0; i < declaredParams.length; i++) {
		if (declaredParams[i].name === param.name) {
			declaredParams.splice(i, 1);
		}
	}
	declaredParams.push(param);
}

function clearParams() {
	declaredParams.splice(0, declaredParams.length);
}

const databaseURL = new require_params_types.InternalExpression("DATABASE_URL", (env) => JSON.parse(env.FIREBASE_CONFIG || "{}")?.databaseURL || "");

const projectID = new require_params_types.InternalExpression("PROJECT_ID", (env) => JSON.parse(env.FIREBASE_CONFIG || "{}")?.projectId || "");

const gcloudProject = new require_params_types.InternalExpression("GCLOUD_PROJECT", (env) => JSON.parse(env.FIREBASE_CONFIG || "{}")?.projectId || "");

const storageBucket = new require_params_types.InternalExpression("STORAGE_BUCKET", (env) => JSON.parse(env.FIREBASE_CONFIG || "{}")?.storageBucket || "");

function defineSecret(name, options = {}) {
	const param = new require_params_types.SecretParam(name, options);
	registerParam(param);
	return param;
}

function defineJsonSecret(name, options = {}) {
	const param = new require_params_types.JsonSecretParam(name, options);
	registerParam(param);
	return param;
}

function defineString(name, options = {}) {
	const param = new require_params_types.StringParam(name, options);
	registerParam(param);
	return param;
}

function defineBoolean(name, options = {}) {
	const param = new require_params_types.BooleanParam(name, options);
	registerParam(param);
	return param;
}

function defineInt(name, options = {}) {
	const param = new require_params_types.IntParam(name, options);
	registerParam(param);
	return param;
}

function defineFloat(name, options = {}) {
	const param = new require_params_types.FloatParam(name, options);
	registerParam(param);
	return param;
}

function defineList(name, options = {}) {
	const param = new require_params_types.ListParam(name, options);
	registerParam(param);
	return param;
}

function expr(strings, ...values) {
	return new require_params_types.InterpolationExpression(strings, values);
}

exports.BUCKET_PICKER = require_params_types.BUCKET_PICKER;
exports.Expression = require_params_types.Expression;
exports.clearParams = clearParams;
exports.databaseURL = databaseURL;
exports.declaredParams = declaredParams;
exports.defineBoolean = defineBoolean;
exports.defineFloat = defineFloat;
exports.defineInt = defineInt;
exports.defineJsonSecret = defineJsonSecret;
exports.defineList = defineList;
exports.defineSecret = defineSecret;
exports.defineString = defineString;
exports.expr = expr;
exports.gcloudProject = gcloudProject;
exports.multiSelect = require_params_types.multiSelect;
Object.defineProperty(exports, 'params_exports', {
  enumerable: true,
  get: function () {
    return params_exports;
  }
});
exports.projectID = projectID;
exports.select = require_params_types.select;
exports.storageBucket = storageBucket;
